### PR TITLE
:sparkles:  auto-approve kubelet serving CSRs

### PR DESCRIPTION
**Pull request by arianvp**
**What this PR does / why we need it**:

This watches CSRs in the target cluster, checks if their DNSNames
and IPAddresses match up with the cluster's Machine, and then approves
the serving certificate

This implements (parts of) https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210222-kubelet-authentication.md#serving-csr-handling

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Implements (parts of) https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210222-kubelet-authentication.md#serving-csr-handling

**Still TODO**

* ship a template where `rotate-server-certificates` is set on `kubelet` and `kubelet-certificate-authority` is set on `apiserver` so that the CSRs are created and the apiserver can talk to the kubelets after they get their certificate signed
* tests
*  Some TODOs still in the code

### CODE REVIEW

**Comment by enxebre** [comment]
Thanks for the giving this a go @arianvp! before proceeding any further with the impl details let's make sure there's community alignment on the short/medium/long term view for the proposal. For that I'd have some expectations:

1 - This feature should be optional and composable and existing running approvers and workflows should not be disrupted. We should not invalidate existing more restrictive approval workflows.

2 - We'd need to clarify, agree layout and document the overall workflow for all possible scenarios e.g client cert, serving cert, bootstrapping, renewal...
E.g Reference impl approvers workflow:
There’s two approvers in the cluster:
    1 - The one given by kube controller manager
    2 - The machine approver.
There is no possible race between them for denial. To prevent conflicts with other approvers If a CSR does not meet criteria the approvers don’t explicitly deny CSRs.
Bootstrapping:
- Kubelet TLS bootstrapping is configured for requesting both client and serving certificates.
- Kube controller manager has no permissions to approve client/serving CSRs for new nodes.
- So it’s up to the machine approver to approve client/serving CSRs for new nodes. If they don meet criteria manual approval is needed.
Rotation:
- Kubelet is configured for rotating both client and serving certificates.
- Kube controller manager approves kubelet renewal client CSRs via system-bootstrap-node-renewal ClusterRoleBinding
- Kube controller manager does not support approval of kubelet renewal serving CSRs.
- So it’s up to the machine approver to approve renewal serving CSRs. If they don meet criteria manual approval is needed.

3 - Also we'd need to specify that as it is, this is not providing node attestation but rather an optional best effort heuristics to validate a Node is legit, e.g as part of the checks we could also have timeouts for CSRs, otherwise as a cluster user with bad intentions I could e.g force triggering autoscaling with aggressive workflows and eventually impersonate a Node by auto generating CSRs.

File: internal/controllers/machine/machine_controller_kubelet_serving_csr.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		for i, org := range cr.Subject.OrganizationalUnit {
+			if org != csr.Spec.Groups[i] {
+				return ctrl.Result{}, fmt.Errorf("csr Groups didn't match O")
+			}
+		}
+
+		condition := certificatesv1.CertificateSigningRequestCondition{
+			LastUpdateTime: metav1.Time{Time: time.Now()},
+		}
+		if err := validateCSR(cr, machine.Status.Addresses); err != nil {
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_kubelet_serving_csr.go:92]: could use conditions.Mark* library and this should probably compute for Machine readiness Summary.

File: internal/controllers/machine/machine_controller_kubelet_serving_csr.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		} else {
+			condition.Type = certificatesv1.CertificateApproved
+			condition.Reason = "CSRValidationSucceed"
+			condition.Status = "True"
+			condition.Message = "Validation was successful"
+
+		}
+		csr.Status.Conditions = append(csr.Status.Conditions, condition)
+
+		if _, err := kubeClient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, &csr, metav1.UpdateOptions{}); err != nil {
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_kubelet_serving_csr.go:107]: This should be somehow composable. We should not break existing workflows and preclude existing ad-hoc approvers from being more restrictive, by approving explicitly here we are effectively defeating any existing approval workflow.

File: internal/controllers/machine/machine_controller_kubelet_serving_csr.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+// validateCSR checks that all IPAddresses and DNSNames in csr are what we expect from the Machine definition and that the CSR is part of NodesGroup
+func validateCSR(csr *x509.CertificateRequest, addresses []clusterv1.MachineAddress) error {
+	containsGroup := false
+
+	for _, group := range csr.Subject.OrganizationalUnit {
+		if group == NodesGroup {
+			containsGroup = true
+			break
+		}
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_kubelet_serving_csr.go:123]: all this logic would need to be uni tested.

**Comment by arianvp** [comment]
Thanks for the detailed response! Happy to await the discussion and looking forward to fleshing this idea out further.  In the meantime I will probably keep messing a bit with this fork (And take some of the suggestions you left) as this is the first time I'm implementing a k8s controller. 

One thing that might be useful is to have a list of existing approvers in existing infrastructure providers 

I found at least one:

* [cluster-api-provider-hetzner](https://github.com/syself/cluster-api-provider-hetzner/blob/main/controllers/csr_controller.go) which the discussion with the authors on Slack sparked me working on this PR. They're happy to move to an upstream approver but we shouldn't break their existing flow.

**Comment by enxebre** [comment]
>One thing that might be useful is to have a list of existing approvers in existing infrastructure providers so that we know what workflows we should not break.

Sounds like a good exercise though to clarify, it's not only about existing ones but also upcoming ones. So regardless of the existing workflows I think that by design we should preserve the ability for consumers to plug in their own approval workflow if they choose to, i.e this should be composable.

File: internal/controllers/machine/machine_controller_kubelet_serving_csr.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	restConfig, err := remote.RESTConfig(ctx, controllerName, r.Client, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+
+	node := corev1.Node{}
+	if err := remoteClient.Get(ctx, client.ObjectKey{Name: nodeRef.Name}, &node); err != nil {
+		return ctrl.Result{}, err
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_kubelet_serving_csr.go:53]: this logic in general could be modelled against meaningful abstractions of the domain, e.g getServingCert(), getClientCert(), validateServingCert() which can then unit test and change impl transparently.

**Comment by fabriziopandini** [comment]
I agree with @enxebre we should discuss this and eventually agree on a limited scope/a limited set of workflows to get this moving.
In the meantime only one comment from my side: If I remember well one of the key elements of the entire proposal was the idea of relying on a strong proof for the machine identity (e.g TPM) before approving signing requests. Are we fulfilling this requirement?

**Comment by arianvp** [comment]
> In the meantime only one comment from my side: If I remember well one of the key elements of the entire proposal was the idea of relying on a strong proof for the machine identity (e.g TPM) before approving signing requests. Are we fulfilling this requirement?

I interpreted the proposal as requiring strong proof for the `kubelet-client` certificate  but not for `kubelet-serving`.  I don't see how TPM attestation would help with improving the security of the `kubelet-serving` certificate as address and hostname is something that is external to the machine itself and not part of its cryptographic identity.

Note that the value of `Machine.Status.Addresses` is trusted as it's determined by the infrastructure provider.   We merely check whether the `Node` is not trying to lie about being more than the `Machine` on which it was scheduled.

The TPM attestation will not help you determine whether an IP address or nodeName is valid for a `Node`. That's something that can only be externalised to the infrastructure provider.   This is similar to how Letsencrypt ACME works in a way.

Note that the current situation is that `apiserver` does not check the `kubelet-serving` certificate _at all_ and just blindly trusts the self-signed cert.  Having checked that the `kubelet-serving` certificate is only serving on IP addresses and DNSNames that are defined in the `Machine` is already a way better security posture.

**Comment by arianvp** [comment]
Re-reading the proposal; it seems that indeed TPM might be in the loop for serving. Though i'm not fully clear how.

and `cluster-x-k8s-io/kubelet-serving-tpm`signer is used instead of the built-in `kubernetes.io/kubelet-serving` signer in kube-controller-manager

But I am a bit confused as it also mentions that `--rotate-server-certificates/ServerTLSBootstrap ` should be enabled on the kubelet. but that _always_ requests `kubernetes.io/kubelet-serving` certificates as far as I know and is not configurable. 

Maybe we should revisit / rephrase the proposal first and clarify the situation about the `serving` certificates as how it's currently written down it's a bit confusing to me and it can be interpreted both ways (where one way seems unimplementable to me given kubelet can not be customised to change what type of CSR it requests for serving)

**Comment by arianvp** [comment]
Openshift seems to have a CSR approver on top of cluster-API too https://github.com/openshift/cluster-machine-approver#node-server-csr-approval-workflow

It sets some requirements that we might want to mimmick:

> Requirements for Cluster API Providers
> As discussed in previous sections, cluster-machine-approver imposes some requirements on each Cluster API provider used with the [machine-api-operator](https://github.com/openshift/machine-api-operator). This section serves as a summary of those requirements.

> A Machine must have a NodeInternalDNS set in Status.Addresses that matches the name of the Node. The NodeInternalDNS entry must be present, even before the Node resource is created.
> A Machine must also have matching NodeInternalDNS, NodeExternalDNS, NodeHostName, NodeInternalIP, and NodeExternalIP addresses as those listed on the Node resource. All of these addresses are placed in the CSR and are validated against the addresses on the Machine object.

**Comment by arianvp** [comment]
Hey @yastij just a friendly bump. You mentioned in Slack some time ago that you wanted to share some thoughts on this and I'm curious about them before I move forward

**Comment by yastij** [comment]
once you've done the client certificate attestation, we pretty much know that the process doing the join is who it says it is, so the serving certs wouldn't need that. `cloud-provider-gcp` is following pretty much the same pattern

File: internal/controllers/machine/machine_controller_kubelet_serving_csr.go (hunk 1 above)
Comment (chuckha) [reply internal/controllers/machine/machine_controller_kubelet_serving_csr.go:89]: To clarify, you are suggesting, in addition to setting the CSR condition, to also set the Machine condition using the `conditions.Mark*` functions as well as add this new condition to the machine_controller's SetSummary call?

**Comment by arianvp** [comment]
Okay great. So my assumption that for the `serving` part we do not need to deal with TPM stuff is correct then?

How do we want to move this forward? To me it sounds there are no blockers now except for cleaning up the PR as it works as advertised and implements the security model from the RFC.

**Comment by yastij** [comment]
sounds good to me

**Comment by guettli** [comment]
@arianvp It is sad to see this PR closed. Is there anything which I can do to get it merged?

**Comment by arianvp** [comment]
Somebody would have to pick up where i left off. I unfortunately currently do not have the bandwidth to finish it. Feel free to reopen it and solve the outstanding issues!

**Pull request closed without merging**